internal/streamingnode/server/walmanager/wal_lifetime.go

```diff
@@ -0,0 +1,161 @@
+func newWALLifetime(opener wal.Opener, channel string) *walLifetime {
+	ctx, cancel := context.WithCancel(context.Background())
+	l := &walLifetime{
+		ctx:       ctx,
+		cancel:    cancel,
+		finish:    make(chan struct{}),
+		opener:    opener,
+		statePair: newWALStatePair(),
+		logger:    log.With(zap.String("channel", channel)),
+	}
+	go l.backgroundTask()
+	return l
+}
+
+// walLifetime is the lifetime management of a wal.
+// It promise a wal is keep state consistency in distributed environment.
+// All operation on wal management will be sorted with following rules:
+// (term, available) illuminate the state of wal.
+// term is always increasing, available is always before unavailable in same term, such as:
+// (-1, false) -> (0, true) -> (1, true) -> (2, true) -> (3, false) -> (7, true) -> ...
+type walLifetime struct {
+	ctx    context.Context
+	cancel context.CancelFunc
+
+	finish    chan struct{}
+	opener    wal.Opener
+	statePair *walStatePair
+	logger    *log.MLogger
+}
+
+// GetWAL returns a available wal instance for the channel.
+// Return nil if the wal is not available now.
+func (w *walLifetime) GetWAL() wal.WAL {
+	return w.statePair.GetWAL()
+}
+
+// Open opens a wal instance for the channel on this Manager.
+func (w *walLifetime) Open(ctx context.Context, channel types.PChannelInfo) error {
+	// Set expected WAL state to available at given term.
+	expected := newAvailableExpectedState(ctx, channel)
+	if !w.statePair.SetExpectedState(expected) {
+		return status.NewUnmatchedChannelTerm("expired term, cannot change expected state for open")
+	}
+
+	// Wait until the WAL state is ready or term expired or error occurs.
+	return w.statePair.WaitCurrentStateReachExpected(ctx, expected)
+}
+
+// Remove removes the wal instance for the channel on this Manager.
+func (w *walLifetime) Remove(ctx context.Context, term int64) error {
+	// Set expected WAL state to unavailable at given term.
+	expected := newUnavailableExpectedState(term)
+	if !w.statePair.SetExpectedState(expected) {
+		return status.NewUnmatchedChannelTerm("expired term, cannot change expected state for remove")
+	}
+
+	// Wait until the WAL state is ready or term expired or error occurs.
+	return w.statePair.WaitCurrentStateReachExpected(ctx, expected)
+}
+
+// Close closes the wal lifetime.
+func (w *walLifetime) Close() {
+	// Close all background task.
+	w.cancel()
+	<-w.finish
+
+	// No background task is running now, close current wal if needed.
+	currentState := w.statePair.GetCurrentState()
+	logger := log.With(zap.String("current", toStateString(currentState)))
+	if oldWAL := currentState.GetWAL(); oldWAL != nil {
+		oldWAL.Close()
+		logger.Info("close current term wal done at wal life time close")
+	}
+	logger.Info("wal lifetime closed")
+}
+
+// backgroundTask is the background task for wal manager.
+// wal open/close operation is executed in background task with single goroutine.
+func (w *walLifetime) backgroundTask() {
```

Comment:
Suggest using one background task to process life time of all WALs that is goruntines effective